src/object.jl

```diff
@@ -47,43 +50,199 @@ function Base.unsafe_copyto!(dest::AbstractVector{UInt8}, doff::Integer, src::Ob
     return n
 end
 
-function Base.unsafe_copyto!(dest::Channel, doff::Integer, src::Object, soff::Integer, n::Integer)
-    bytes = getRange(src, soff, n)
-    dest_arr, new_n = take!(dest)
-    new_n == n || throw(ArgumentError("prefetched number of bytes (`$n`) doesn't matched requested number of bytes (`$new_n`)"))
-    copyto!(dest_arr::AbstractVector{UInt8}, doff, bytes)
-    put!(dest, n)
-    return n
+mutable struct TaskCondition
+    cond_wait::Threads.Condition
+    ntasks::Int
+end
+TaskCondition() = TaskCondition(Threads.Condition(), 0)
+mutable struct PrefetchBuffer
+    data::Vector{UInt8}
+    pos::Int
+    len::Int
+end
+Base.bytesavailable(b::PrefetchBuffer) = b.len - b.pos + 1
+
+function _prefetching_task(io)
+    prefetch_size = io.prefetch_size
+    len = io.len
+    ppos = 0
+    download_buffer = Vector{UInt8}(undef, prefetch_size)
+    # Don't allocate the second buffer when we can fit everything to the first one
+    download_buffer_next = len == prefetch_size ? download_buffer : Vector{UInt8}(undef, prefetch_size)
+
+    try
+        while len > ppos
+            n = min(len - ppos, prefetch_size)
+            off = 0
+            # `partition` plays nicely with 1-based ranges, but we need zero based ranges for
+            # `contentRange`
+            rngs = Iterators.partition(ppos+1:ppos+n, io.prefetch_multipart_size)
+            io.cond.ntasks = length(rngs)
+            for rng in rngs
+                put!(io.download_queue, (off, rng .- 1, download_buffer))
+                off += length(rng)
+            end
+            Base.@lock io.cond.cond_wait begin
+                while true
+                    io.cond.ntasks == 0 && break
+                    wait(io.cond.cond_wait)
+                end
```

Comment:
Sorry for the comments all over the place. GitHub is hard.
-----

This confused me at first. the following would be more clear i think:
```suggestion
                while io.cond.ntasks != 0
                    wait(io.cond.cond_wait)
                end
```
or even better:
```suggestion
                while io.cond.ntasks > 0
                    wait(io.cond.cond_wait)
                end
```